From patchwork Fri Sep  7 11:59:47 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wei Yang <richard.weiyang@gmail.com>
X-Patchwork-Id: 10592117
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id F3529139B
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri,  7 Sep 2018 12:00:44 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id D3B2D2AEEC
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri,  7 Sep 2018 12:00:44 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id C806B2AFBA; Fri,  7 Sep 2018 12:00:44 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 777622AEEC
	for <patchwork-kvm@patchwork.kernel.org>;
 Fri,  7 Sep 2018 12:00:44 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728146AbeIGQkc (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Fri, 7 Sep 2018 12:40:32 -0400
Received: from mail-pl1-f196.google.com ([209.85.214.196]:33273 "EHLO
        mail-pl1-f196.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727820AbeIGQkc (ORCPT <rfc822;kvm@vger.kernel.org>);
        Fri, 7 Sep 2018 12:40:32 -0400
Received: by mail-pl1-f196.google.com with SMTP id b30-v6so4227887pla.0
        for <kvm@vger.kernel.org>; Fri, 07 Sep 2018 04:59:55 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id;
        bh=R/CnLYBoEQ8ebQ5BtYs3AL5IKqUYBGvlXPd1MwbB7bo=;
        b=hTygRTU906v3LxExZ6dXA2G4zC4JeTkl4MVdRc4xVTTJVjAzPhOie1/voM1CSWtxN+
         ftHKtnLUKGMFT8a22tx3CRwX0lGHIcQ2HC7K+sFEJN14ehDHE47Fpzh8C6QZLyXxMu5E
         jz+e6kFEtFkUs35dyF66D0HkMyyWlHFGzgoRwqGMQBvAELPeJA8qD2GieRkeMKAOK3kL
         E0im6ohvDjl70URBbCteZ5rYjuF/Cm3igWr6+lvxOzIZWH3AiFFLtrbn4+9LGSnXHrjJ
         s5mtzqORL5CWBqg3FTh7GSVAotpsM/BCXtZtkPU//+zRoHqeX4CdSu6I6w06KQ8/WWLb
         OZtA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id;
        bh=R/CnLYBoEQ8ebQ5BtYs3AL5IKqUYBGvlXPd1MwbB7bo=;
        b=ctQg29SCwFXyvmf4ZKRchXZ2uyZvq9v5OCrR2nuc5cjPBUYcGWZjtXnlWb7hm4+V3A
         u7qZKxEvK478/pTX7ByFRHXt88wUPebNCd/tkD8+VKlkuLHp25ZurdVE/mrNURTCSWkP
         Kp+umAmJTKCSefZCp3hYwrxhx0T4Qf6a2Id0aYyApUM0wv7oYKCQY4SzsHjfzACHPd9n
         uQ64LZHo0IFRPz9V1E6HL8Tx3v1joGmMTJY4dOHi0nXmIugjzYLaAnW8ixlVPucTAYlo
         WORU9Rqo9Vv1Ic2N7zh3k7+AOdeBQY21fOdD5n5Vy2giA173nMHQNHSed4F7MufVSdq5
         jrkA==
X-Gm-Message-State: APzg51BPJkNW7cozxaL8wJmOCMun/Sdf6QwjiAtWTb7xndn011Veuioz
        SzG4yQuhxG6x6W8Us88NtHU=
X-Google-Smtp-Source: 
 ANB0VdaTFfptaCr6vNHkiI/CEeezpht7wZMfqC57Y4C5mHJdUrrnvUfio7JyWO+MYm9Ey1YRy9E3bA==
X-Received: by 2002:a17:902:82c5:: with SMTP id
 u5-v6mr7522228plz.83.1536321595032;
        Fri, 07 Sep 2018 04:59:55 -0700 (PDT)
Received: from localhost ([185.92.221.13])
        by smtp.gmail.com with ESMTPSA id
 o10-v6sm8560337pgp.70.2018.09.07.04.59.53
        (version=TLS1_2 cipher=ECDHE-RSA-AES128-GCM-SHA256 bits=128/128);
        Fri, 07 Sep 2018 04:59:54 -0700 (PDT)
From: Wei Yang <richard.weiyang@gmail.com>
To: pbonzini@redhat.com, rkrcmar@redhat.com, liran.alon@oracle.com,
        sean.j.christopherson@intel.com
Cc: tglx@linutronix.de, mingo@redhat.com, x86@kernel.org,
        kvm@vger.kernel.org, Wei Yang <richard.weiyang@gmail.com>
Subject: [PATCH v3] KVM: x86: don't reset root in kvm_mmu_setup()
Date: Fri,  7 Sep 2018 19:59:47 +0800
Message-Id: <20180907115947.1947-1-richard.weiyang@gmail.com>
X-Mailer: git-send-email 2.15.1
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

Here is the code path which shows kvm_mmu_setup() is invoked after
kvm_mmu_create(). Since kvm_mmu_setup() is only invoked in this code path,
this means the root_hpa and prev_roots are guaranteed to be invalid. And
it is not necessary to reset it again.

    kvm_vm_ioctl_create_vcpu()
        kvm_arch_vcpu_create()
            vmx_create_vcpu()
                kvm_vcpu_init()
                    kvm_arch_vcpu_init()
                        kvm_mmu_create()
        kvm_arch_vcpu_setup()
            kvm_mmu_setup()
                kvm_init_mmu()

This patch set reset_roots to false in kmv_mmu_setup().

Signed-off-by: Wei Yang <richard.weiyang@gmail.com>
Reviewed-by: Liran Alon <liran.alon@oracle.com>
---
v3: add a comment in code based on Liran comment.
v2: adjust changelog based on Sean Christopherson comment
---
 arch/x86/kvm/mmu.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/mmu.c b/arch/x86/kvm/mmu.c
index b2b88454b1d8..899c029cff0d 100644
--- a/arch/x86/kvm/mmu.c
+++ b/arch/x86/kvm/mmu.c
@@ -5413,7 +5413,12 @@ void kvm_mmu_setup(struct kvm_vcpu *vcpu)
 {
 	MMU_WARN_ON(VALID_PAGE(vcpu->arch.mmu.root_hpa));
 
-	kvm_init_mmu(vcpu, true);
+	/*
+	 * kvm_mmu_setup() is called only on vCPU initialization.  
+	 * Therefore, no need to reset mmu roots as they are not yet
+	 * initialized.
+	 */
+	kvm_init_mmu(vcpu, false);
 }
 
 static void kvm_mmu_invalidate_zap_pages_in_memslot(struct kvm *kvm,
